Pressingly/lago-api
# Feat/pinet 174 event for consumption activity

# Motivation / Background

When publisher core user action is authorized by GetLago+, a consumption event need to  be emitted to record this user action in GetLago https://thepressingly.atlassian.net/browse/PINET-174

*This Pull Request has been created because:*
Emit consumption event when authorization success

# Detail

*This Pull Request changes:*
- Add `create_get_lago_event` to successful authorization flow
- Create `EmitService` to handle event emitting logic

# Additional information

### How to test
- Run this command which is similar with https://github.com/Pressingly/lago-api/pull/10
```shell
curl --location --request POST "http://localhost:3000/v1/entitlement/authorize" \
  --header "Authorization: Bearer GetLago_API_token" \
  --header 'Content-Type: application/json' \
  --data-raw '{
  "userId": "GetLago_customer_id",
  "publisherId": "Publisher_id_we_create_for_the_publisher_when_they_are_onboarded",
  "actionName": "read",
  "context": {},
  "resource": {
    "id": "your_plan_id_in_previous_step",
    "name": "Liverpool is winning Champion League",
    "type": "article",
    "author": "author name",
    "tags": [
      "Climate change"
    ],
    "category": "basketball"
  },
  "timestamp": "2022-03-01T12:34:56+02:00"
}'
```
- Visit `http://localhost:8080/developers/debugger` when running the system on your local. This is FE repo https://github.com/Pressingly/admin-publisher-portal. Click on "Event" tab to check the latest event, you should see something like this
<img width="1207" alt="image" src="https://github.com/Pressingly/lago-api/assets/33661599/137ff845-e4f1-4d98-aa1d-fc46ba4f4691">

- Visit OpenAPI at this endpoint `http://localhost:3000/api-docs/index.html` when running at local
<img width="1410" alt="image" src="https://github.com/Pressingly/lago-api/assets/33661599/9703068c-c340-4d46-a964-dfc4dfd2f4c6">

*TIP: Provide additional information such as screenshots, benchmarks, reference to other repositories or alternative solutions*

# Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
